common/numerated/src/iterators.rs

```diff
@@ -0,0 +1,330 @@
+            Err(NewWithLenError::OutOfBounds(err)) => Err(err),
+        }
+    }
+}
+
+impl<T: Display> Display for IntervalIterator<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        if let Some(interval) = &self.0 {
+            write!(f, "{}", interval)
+        } else {
+            write!(f, "∅")
+        }
+    }
+}
+
+/// Helper struct to iterate over intervals from `tree1`, which are not in `tree2`.
+///
+/// See also [`IntervalsTree::difference`].
+pub struct DifferenceIterator<T: Numerated, I: Iterator<Item = Interval<T>>> {
+    pub(crate) iter1: I,
+    pub(crate) iter2: I,
+    pub(crate) interval1: Option<Interval<T>>,
+    pub(crate) interval2: Option<Interval<T>>,
+}
+
+impl<T: Numerated, I: Iterator<Item = Interval<T>>> Iterator for DifferenceIterator<T, I> {
+    type Item = Interval<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let interval1 = if let Some(interval1) = self.interval1 {
+                interval1
+            } else {
+                let interval1 = self.iter1.next()?;
+                self.interval1 = Some(interval1);
+                interval1
+            };
+
+            let interval2 = if let Some(interval2) = self.interval2 {
+                interval2
+            } else if let Some(interval2) = self.iter2.next() {
+                interval2
+            } else {
+                self.interval1 = None;
+                return Some(interval1);
+            };
+
+            if interval2.end() < interval1.start() {
+                self.interval2 = None;
+                continue;
+            }
+
+            self.interval2 = Some(interval2);
+
+            if interval1.end() < interval2.start() {
+                self.interval1 = None;
+                return Some(interval1);
+            } else {
+                if let Some(new_start) = interval2.end().inc_if_lt(interval1.end()) {
+                    self.interval1 = Interval::new(new_start, interval1.end());
+                    debug_assert!(self.interval1.is_some(), "`T: Numerated` impl error");
+                } else if interval1.end() == interval2.end() {
+                    self.interval1 = None;
+                    self.interval2 = None;
+                } else {
+                    self.interval1 = None;
+                }
+
+                if let Some(new_end) = interval2.start().dec_if_gt(interval1.start()) {
+                    let res = Interval::new(interval1.start(), new_end);
+                    debug_assert!(res.is_some(), "`T: Numerated` impl error");
+                    return res;
+                } else {
+                    continue;
+                }
+            }
+        }
+    }
+}
+
+/// Helper struct to iterate over voids in tree.
+///
+/// See also [`IntervalsTree::voids`].
+pub struct VoidsIterator<T: Numerated, I: Iterator<Item = Interval<T>>> {
+    pub(crate) inner: Option<(I, Interval<T>)>,
```

Comment:
Shouldn't it be different Options or even only optional interval?

Reply:
What the reason to make not like this ? It's the most simple solution